Multi threaded RDB Save

Nicky-2000 opened the pull request:
# Multi-threaded RDB Save

### Contents
1. Introduction 
2. Performance Improvement using RDB Threads
3. Profiling Current, Single-Threaded, Implementation
4. Detailed Design
5. Testing
6. RDB Threads Configuration
7. Future Extension

---

### Introduction:
This PR introduces a multi-threaded RDB Save, which aims to improve Valkey's persistence performance. There will be a future PR that introduces multithread RDB Load.

**Motivation for PR:**
This PR addresses Issue: https://github.com/valkey-io/valkey/issues/1483

Currently, RDB operations, including snapshots and full synchronizations, are single-threaded and often take a very long time (especially with large datasets). This poor performance can lead to Out-of-Memory (OOM) errors due to Copy-On-Write (COW) overhead, long recovery times, reduced availability and failures during cluster scaling events. Host machines running Valkey often have underutilized CPU cores, and since the data encoding and compression work during RDB saves is highly repetitive, it's an ideal situation for to use threading to improve performance.

TODO: 
1. Remove min and max value from rdb thread flush buffer threshold as per Jun's suggestion. 
---

# Performance Improvements
### Test Environment:
Performance tests were run on a Google Cloud n2-highmem-48 VM with the following hardware configuration:
- **Machine Type:** n2-highmem-48
- **CPU:** 48 vCPUs (24 physical cores) on an Intel Cascade Lake platform.
- **Memory:** 384 GB of RAM.
- **Storage:** A persistent 1 TB SSD was attached for RDB file saves.
- **Write Bandwidth (Important)**: ~ 760 MB/s


 
### Workloads Tested
It is important to note that the test data (keys and values) used for these benchmarks was generated with valkey-benchmark and is entirely non-compressible.
| Number of Keys | Value Size | Notes |
|:---:|:---:|:---|
| 2 Billion | 50 B | Small Value Size Workload. |
| 1 Billion | 300 B | Small Value Size Workload. |
| 500 Million | 500 B | Small Value Size Workload. |
| 50 Million | 5 KB | Medium Value Size Workload. |
| 10 Million | 20 KB | Medium Value Size Workload. |
| 5 Million | 64 KB | Slightly smaller than 64KiB buffer write threshold. |
| 5 Million | 66 KB | Slightly larger than 64KiB buffer write threshold. |
| 1 Million | 100 KB | Large Value Size Workload. |
| 0.5 Million | 262 KB | Slightly smaller than 256KiB max buffer capacity threshold.|
| 0.5 Million | 272 KB | Slightly larger than 256KiB max buffer capacity threshold.|
| 0.1 Million | 1000 KB | Large Value Size Workload. |

All tests were run using the `run_full_save_benchmark.sh` test script located in the [valkey-rdb-benchmarking]( https://github.com/Nicky-2000/valkey-rdb-benchmarking). This bash script is a wrapper of [save_benchmark.py](https://github.com/Nicky-2000/valkey-rdb-benchmarking/blob/main/scripts/save_benchmark.py) python script with different arguments passed in (different workload configurations).

The code for the graphs and the raw Data can be found here: [Visualization Code and Raw Data Notebook]( https://github.com/Nicky-2000/valkey-rdb-benchmarking/blob/main/notebooks/large_tests_64k_buffer/performance_improvement_vs_value_size.ipynb)

### Key Metrics:
Speed up (num_threads) = (Single Thread Save Time) / (Save Time with 'num_threads')
Efficiency (num_threads) = (Speedup with 'num_threads') / ('num_threads')
   
### Results:
Multithreading significantly speeds up the runtime of RDB Save, especially for workloads with "small" keys/values or if compression is on. For large keys, and compression off, there are limited improvements. For the 66KB Value Size Workload there was a 7% Regression when the number of threads is larger than 3.

#### Speed Up and Efficiency For Small Value Size Workloads (50 - 500 Bytes)
##### Compression OFF
With compression off, we observed a substantial speedup for small value size workloads. A 50 B value size workload showed a speedup of over 3x with 4 threads. Workloads with value sizes of 300 B and 500 B achieved a speedup of 3x to 3.2x.

<img width="1334" height="649" alt="Screenshot 2025-08-08 1 58 23 PM" src="https://github.com/user-attachments/assets/bd40729c-ce42-4588-90cf-b1169fe5f849" />

##### Compression ON
With compression on, the performance gains are even larger. For a very small value size (50 B), we achieved up to a 10x speedup with 10 threads. For value sizes in the 300-500 B range, a 4.5x speedup was observed.
<img width="1337" height="660" alt="Screenshot 2025-08-08 2 01 30 PM" src="https://github.com/user-attachments/assets/18805477-637f-40bd-8fac-8305222a2c0b" />


#### Speed Up and Efficiency For Medium Value Size Workloads (5 KB - 66 KB)

##### Compression OFF
For medium-sized keys with compression off, the improvement was minimal, ranging from 1x to 1.2x speedup. This is because the keys are large enough to cause the operating system to buffer them effectively, leading to disk I/O throughput becoming the primary bottleneck.
<img width="1338" height="664" alt="Screenshot 2025-08-08 1 59 50 PM" src="https://github.com/user-attachments/assets/124a67d2-341e-4784-8af3-02109a1818ed" />

##### Compression ON
Enabling compression for medium value size workloads yielded a substantial speedup of 3.5x to 4x.
<img width="1338" height="661" alt="Screenshot 2025-08-08 2 02 02 PM" src="https://github.com/user-attachments/assets/5b25eb9d-14a3-47b2-ab2e-2a69dda026f5" />

#### Speed Up and Efficiency For Large Value Size Workloads (100 KB - 1 MB)

##### Compression OFF
Similar to medium value size workloads, we observed limited improvement of 1x to 1.2x speedup for large value size workloads with compression off. The system remains bottlenecked by disk I/O.
<img width="1342" height="671" alt="Screenshot 2025-08-08 2 00 30 PM" src="https://github.com/user-attachments/assets/9143bb02-c52b-4aaf-874a-38ef47645399" />

##### Compression ON
With compression on, large value size workloads demonstrated a significant speedup of approximately 4x.
<img width="1336" height="661" alt="Screenshot 2025-08-08 2 02 44 PM" src="https://github.com/user-attachments/assets/c1244c0f-e2fc-4157-b450-9ee972a077bd" />


### Raw Runtime comparison with Compression On and Off
It is important to note that the values used in our tests were generated with valkey-benchmark and were **_entirely non-compressible_**. This resulted in the RDB files for both compression ON and OFF being identical in size.

Therefore, the multi-threaded speedup shown in these results is due solely to the effect of parallelizing the CPU intensive operations , without any additional benefit from a smaller file size. If the data had been compressible, we would expect the compression ON runs to be even faster due to less bytes needing to be written to persistent storage. This provides a strong motivation to evaluate enabling RDB compression in environments using multi-threaded RDB save. Furthermore, because we can distribute the compression workload across multiple threads, there could be significant value in using a more powerful compression algorithm (with a higher compression ratio). This would not only makes the saved file sizes smaller but could also reduce the overall save runtime, since fewer bytes will need to be written to storage. See the "Testing on More Complex, Real World Workloads" and "Performance and Compression Ratio Relationship" sections for more details on this.
  
#### Improvement on Small  Value Size Workloads
<img width="2734" height="1019" alt="Screenshot 2025-08-08 2 03 59 PM" src="https://github.com/user-attachments/assets/7c9448f9-ba9d-40fe-80b2-b359fe056185" />

#### Improvement on Medium Key Size Workloads
<img width="2734" height="1019" alt="Screenshot 2025-08-08 2 04 41 PM" src="https://github.com/user-attachments/assets/315ef011-3095-483b-8796-6fd72a69bcc9" />

#### Improvement on Large Value Size Workloads
<img width="2734" height="1014" alt="Screenshot 2025-08-08 2 05 12 PM" src="https://github.com/user-attachments/assets/ec4cec2f-c3aa-4825-bdaf-464f5b4c1ea6" />

### Testing on More Complex, Real World Workloads: 
The performance tests described above used uncompressible string values of different lengths. Therefore, there was no change in the generated RDB file size if compression was on.

Valkey, however, supports more complex data types such as lists, sets, and JSON. Performance was also evaluated using these types, and a general trend emerged: the more complex the data type, the greater the potential performance improvement.

For example, the figure below shows the performance gains when saving a database filled with large, complex JSON objects. While the specific speedup depends on the nature of the data, the testing shows it's possible to achieve a 20x+ speedup when compression is enabled, and a 4x to 10x speedup with compression off. Depending on the workload the speedup could be even greater.

The JSON Generation function used to create values for this workload [can be found here](https://github.com/Nicky-2000/valkey-rdb-benchmarking/blob/d92deab821bb86240aa77b7e19ddfbb5920f58df/utilities/key_value_generation_utilities.py#L141).

<img width="1299" height="411" alt="Screenshot 2025-08-28 3 54 20 PM" src="https://github.com/user-attachments/assets/090809bd-66cc-46ad-bccf-16358cc4b4d9" />


### Performance and Compression Ratio Relationship
The performance test on the complex JSON (above) showed that the ratio between the fastest save time with compression off and the fastest save time with compression on is approximately equal to the compression ratio (the original file size divided by the compressed file size)!

This strongly motivates the value of introducing a better-performing compression algorithm (with a  higher compression ratio) into Valkey as it could lead to faster overall RDB Save times (and Load).

### Important Note on Regression:
The multi-threaded implementation showed a performance regression when processing key-value pairs with a 66KB value size. This was the only workload where the multi-threaded approach was slower than the single-threaded version.

The cause of this regression is the interaction between the multi-threaded write buffer and the data size. The implementation uses a write buffer with a 64KB threshold. Since the 66KB value size causes the buffer to exceed this limit, a write operation is immediately triggered for each key/value pair. This means we are writing to the disk after buffering exactly one key-value pair. Therefore we lose much of the performance benefits of multi-threading. Instead, for this workload, adding additional threads create a performance penalty due to coordination overhead. To help alleviate this issue you can change the buffer write threshold. See the "RDB File Writing Mechanism" section for more details on this buffer. 

The regression becomes noticeable at 4 threads and worsens as more threads are added, reaching a peak of a 7% slower runtime than the single-threaded baseline when running with 8 threads..
<img width="1342" height="669" alt="Screenshot 2025-08-08 2 06 46 PM" src="https://github.com/user-attachments/assets/15ddaa42-ae89-4efb-83a9-c38385a8197e" />

---

# Profiling Current, Single-Threaded, Implementation
To inform the design of RDB Save multi-thread, performance profiling was conducted to identify the bottlenecks under different workloads.  

### Profiling Environment
Note: This profiling was done on a different machine than the "Performance Improvement Tests"

- **CPU:** 64 vCPUs (32 physical cores) on an AMD EPYC 7B13 platform.
- **Memory:** 117 GiB of RAM.
- **Storage:** A 500 GiB SSD was attached and mounted for RDB file saves and benchmarking.
- **Write Bandwidth (Important):** ~ 402 MB/s

### Results
<img width="814" height="503" alt="Valkey Data Throughput by Value Size and Compression" src="https://github.com/user-attachments/assets/429db5d5-5602-4287-819c-a1202a159c94" />

The graph above shows that for compression OFF we are I/O bound when keys/values are greater than ~5,000 - 10,000 Bytes (for this system). When keys/values are  < ~5000 Bytes then we are CPU Bound . When compression is ON we are always CPU bound. 

This shows that there are significant performance gains to be made by parallelizing the CPU intensive work of encoding (and compressing) the data.  

---

# Detailed Design
### High-Level Architecture
The proposed architecture uses a multi-threaded encode, serial-write model. This design separates the Valkey RDB Save process into two phases:

1. Parallel Encoding: Multiple worker threads concurrently encode Valkey objects and compress the data (if enabled). Each thread has a local buffer to store its encoded data, which avoids contention during this phase. The local buffer is bounded in size to reduce the probability of OOM errors.

2. Serial Writing: Upon filling up its local buffer with data, each thread acquires a shared lock before writing data to a single RDB file. Therefore, only a single thread writes to the file at any point in time. This phase is limited by the disk's I/O throughput and is the new bottleneck after the encoding is parallelized.


### Work Allocation
Work is partitioned at the **Bucket level**, which provides an ideal granularity for division and is compatible with both Cluster and Standalone modes. Note: We partition and process each hashtable. So in cluster mode, when there are multiple slots (I.e hashtables), we will partition and process each of them sequentially.

To handle tables in the middle of a `rehash` operation, threads are assigned a `start_index` and a `stride` to iterate through a flattened view of "Live-Buckets." This design ensures all data is processed correctly while the rehashing is paused.

The diagram below shows how threads will be allocated certain buckets in the the "flattened" (logical) view of the tables[0] and tables[1] bucket arrays. Note: this is a view of the "top level buckets". All child buckets will be processed by the thread assigned to the top level bucket that is the ancestor of the said child bucket. 

![RDB Save Multi-Thread Diagrams (31)](https://github.com/user-attachments/assets/409c46f2-63cb-40e3-8388-b4afea566fca)


The next diagram shows the full flow of how the main thread allocates work for the worker threads for a single hashtable. Note: In the code the "Main Thread" is also responsible for processing a portion of the hashtable. The diagram does not include this for simplicity. 

![RDB Save Multi-Thread Diagrams (32)](https://github.com/user-attachments/assets/6e7d312b-ef5f-48d9-9029-6be348993eb4)

### RDB File Writing Mechanism

Each worker thread follows a coordinated process:
1.  **Buffer Data**: Threads encode key-value pairs into a local buffer.
2.  **Acquire Write Lock**: When a buffer is full (exceeds a `WORKER_BUFFER_DEFAULT_SIZE` currently set to 64KB), the thread acquires a shared write lock.
3.  **Write**: The thread writes its buffer to the RDB file.
4.  **Release Lock**: The lock is released, and the thread continues processing more keys.

The diagram below shows how each thread processes its allocated buckets, by striding through the hashtable, stores encoded data in a thread local buffer, and dumps data to the rdb file / target Rio. 
![RDB Save Multi-Thread Diagrams (33)](https://github.com/user-attachments/assets/eae71ddb-92bc-4a5d-99f2-c0b86b9cdf11)

### Handling Large Keys/Values
To prevent Out-of-Memory (OOM) errors, a **dynamic buffering strategy with a direct-to-file fallback** is used. If a key-value pair would exceed a hard buffer limit (`WORKER_BUFFER_CAPACITY_LIMIT` currently set to 256KB), the thread flushes its existing buffer and then streams the large key's data directly to the file, bypassing the buffer while holding the write lock. This ensures that each thread is allocated a maximum of 256KB of memory. 

### Threading Model
We chose to abstract the `JobQueue` implementation into `thread_common.c` and create a new set of dedicated functions for RDB thread initialization and management located in `rdb_threads.c`. This approach avoids tight coupling with the existing I/O and BIO threads, allowing for independent control over the number of threads, thread local storage, thread CPU affinity, and job queue sizes.

**Other Options Considered:**
* **I/O Threads**: Rejected due to tight coupling, potential for CPU contention during BGSAVE, and inflexible job queues.
* **BIO Threads**: Rejected because the model's fixed number of threads and "fire-and-forget" mentality are incompatible with the active coordination and variable thread count required for RDB operations.

---

# Testing 
### Unit Tests 
New unit tests were added for the `hashtableStrideNext` function in `test_hashtable.c`. These tests ensure the correct, complete, traversal of the hashtable occurs in the multi-threaded context.

### TCL Integration Tests
A new TCL test configuration flag (`rdb-threads`) was added. This flag configures the Valkey servers to run with rdb-threads active.

The full Valkey TCL test suites were run with the `rdb-threads` flag enabled. All existing tests passed, confirming that this implementation is backward-compatible and introduces no regressions or unintended changes to the RDB format or replication protocol.

Testing command used: ` ./runtest --large-memory --rdb-threads`

I also modified integration/rdb.tcl, integration/replication-2.tcl and integration/replication-3.tcl to run in a loop with different number of rdb-threads enabled.

I am open to feedback on this testing strategy. We could add a "daily" job (like there is for IO threads) that will run the test suite with the --rdb-threads flag. 

---

# RDB Threads Configuration
This PR introduces a new configuration directive, `rdb-threads`, to control the number of threads used for RDB saving.

This directive specifies the number of parallel threads to use for encoding and compressing data during an RDB save operation. The default value is `1`, which reverts to the existing single-threaded behavior.

`rdb-threads <num_threads>`
-  `num_threads`: An integer value in the range 1 to 256
- Behavior: If `num_threads` is set to `N`, the main thread will function as one of the save threads. The system will then spawn `N-1` additional worker threads. For example, if `rdb-threads` is set to `5`, a total of five threads—the main thread and four worker threads—will be used. This is the same behavior as IO threads.

Using a higher number of threads can significantly improve RDB save performance, especially on multi-core systems. The optimal number of threads often depends on the workload and the number of available CPU cores. Analyzing the **_Performance Improvements_** section can help inform your decision about how many threads to use.

---

# Potential Future Extension: Sharded RDB Files
With this threading model in place, we can transition, with not too much effort, to "sharded" rdb files. I.e. we can pass a different target Rio to each thread (or groups of threads). This would enable us to write to multiple RDB Files or multiple sockets on a replica.

**Benefits to sharded RDB files:**
1. This would enable easier multi-threading of RDB Load. Each thread could be assigned one of the sharded RDB files/streams. 
     - Note: The current RDB version (non-sharded) does enable some parallelization of RDB Load by reading "chunks" of the RDB file and then passing them to threads for decoding and insertion into the database. This version of RDB Load (which is a non-breaking backward compatible change) will be done in a future PR.

2. Sharded RDB files would enable concurrent writing to multiple persistent disks. This would help us overcome the I/O bottleneck by writing sharded RDB files to separate disks simultaneously.

**Issues with sharded RDB files:**
1. This would be a significant breaking change to the RDB file format. And would break backwards compatibility.
2. Increased complexity in handling the sharded RDB files.

Nicky-2000 committed:
Abstract out JobQueue:

Change the name of IOJobQueue -> JobQueue and
move it to another file called thread_common.h/c

The reason for this change is to enable the JobQueue
to be reused by RDB Threads.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Create rdb_threads.h/c file.

This includes the thread orchestration and
handler functions for the multi-threaded rdb Save
implementation.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Add hashtableStrideNext iterator and
hook up rdb_threads entrypoint in rdb.c

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Add unit tests for hashtabelStrideNext.

Add rdb-threads flag for TCL integration tests

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Include thread_common.h in files that use
inMainFunction()

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Spellcheck and formatting changes

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Add thread_common.c and rdb_threads.c to cmake
source files.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Fix formatting error

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Ensure that we resume rehashing on the hashtable
being processed if we error out during multi-thread
RDB Save.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Add HASHTABLE_ITER_INCLUDE_IMPORTING to kvstore
iterator in rdb_threads

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Formatting

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

luoos commented on src/hashtable.c:
```diff
@@ -323,6 +323,13 @@ typedef struct {
+typedef struct BucketLocation {
+    uint8_t table_index;
+    size_t bucket_index;
+    uint8_t valid;
```
Can we have a comment explaining what `valid` means?

luoos commented on src/hashtable.c:
```diff
@@ -2075,6 +2085,188 @@ bool hashtableNext(hashtableIterator *iterator, void **elemptr) {
+ * (At this point, B0's entire chain is processed.)
+ *
+ * 2. The NEXT call after B0's chain is exhausted:
+ * - STRIIDE! Jumps 2 logical buckets forward.
```
typo?

luoos commented on src/hashtable.c:
```diff
@@ -2075,6 +2085,188 @@ bool hashtableNext(hashtableIterator *iterator, void **elemptr) {
+ * (At this point, B2's entire chain is processed.)
+ *
+ * 3. The NEXT call after B2's chain is exhausted:
+ * - STRIIDE! Jumps 2 logical buckets forward.
```
typo?

luoos commented on src/rdb_threads.c:
```diff
@@ -0,0 +1,377 @@
+
+/* --------- RDB Worker Threads Core Logic --------- */
+
+static void *RDBThreadMain(void *myid) {
```
As discussed before, the word `Main` is overloaded. Maybe pick a different word?

Nicky-2000 replied:
I think its best to leave it like this since IOThreads use this convention with IOThreadMain. However, if you have a strong preference for something like "RDBThreadEntryPoint" I'll change it.

luoos commented on src/rdb_threads.h:
```diff
@@ -0,0 +1,48 @@
+#include "server.h"
+#include "thread_common.h"
+
+/* Threshold for flushing a worker's buffer to the main RDB file (64 KB). */
```
it doesn't have to be file

luoos commented on src/rdb_threads.h:
```diff
@@ -0,0 +1,48 @@
+
+/* Threshold for flushing a worker's buffer to the main RDB file (64 KB). */
+#define WORKER_BUFFER_DEFAULT_SIZE 64 * (1024)
+/* Maximum capacity for a worker's buffer. Keys causing this limit to be exceeded are streamed directly to RDB file (256 KB). */
```
Maybe explain more about the big key scenario? Why do we need two thresholds, and how do they solve the big key case.

luoos commented on src/rio.c:
```diff
@@ -57,6 +57,102 @@
+                return 0;
+            }
+        }
+        // The caller (rdbEncodeHashtableRange) is responsible for clearing r->io.buf_to_underlying.ptr/pos.
```
nit: inconsistent comment style

luoos reviewed (commented):
Overall LGTM. Some minor comments. Can we have some dedicated tcl for this feature?

dvkashapov commented:
> BIO Threads: Rejected because the model's fixed number of threads and "fire-and-forget" mentality are incompatible with the active coordination and variable thread count required for RDB operations.

I get the fixed num of threads part but can you explain a little what type of coordination is needed for RDB? I've looked a little at the code and do you mean the part where `rdbSaveDbMultiThreaded` waits for all jobs to be finished in `drainRDBThreadsQueue`?

> This approach avoids tight coupling with the existing I/O and BIO threads, allowing for independent control over the number of threads, thread local storage, thread CPU affinity, and job queue sizes.

To avoid introducing third thread type we could first make BJM (#356) come to life first or maybe unify thread types (#2376). Once this is done, multithreaded RDB save could be added as a job type. This way we won't have to create 3rd type of threads that needs to be maintained. Because BJM is still in progress, we can modify it a little for the needs of coordination with the main thread. What do you think?

Nicky-2000 commented:
Thanks for the feedback! Here are my thoughts. 

> I get the fixed num of threads part but can you explain a little what type of coordination is needed for RDB? I've looked a little at the code and do you mean the part where `drainRDBThreadsQueue` waits for all jobs to be finished in `rdbSaveDbMultiThreaded`?

You're right about drainRDBThreadsQueue. That's one of the key differences. The RDB saving process can't use a simple "fire-and-forget" model because it requires active coordination.

We use a fork-join pattern: the main thread allocates (forks) work to the workers (like saving slot-1 in cluster mode), and then we "join" them by waiting for them all to finish. We need this coordination to properly propagate errors (one thread fails means everything should fail). It's also necessary so the main thread knows when it's safe to resume hashtable rehashing.

> To avoid introducing third thread type we could make BJM (#356) come to life first or maybe unify thread types (#2376). Once this is done, multithreaded RDB save could be added as a job type. This way we won't have to create 3rd type of threads that needs to be maintained. Because BJM is still in progress, we can modify it a little for the needs of coordination with the main thread. What do you think?

I agree that a unified threading model is a great long-term goal.

However, there are two key differences that make a dedicated approach more pragmatic for now. First, as discussed, RDB requires a coordinated "fork-join" pattern, which is fundamentally different from BIO's "fire-and-forget" model. Second, RDB threads are ephemeral (they only exist for a specific save/load operation), unlike the permanent daemon threads used for BIO.

Designing a generic manager to handle both patterns and lifecycles is a significant architectural undertaking. This PR takes a direct approach and uses the familiar I/O threads pattern, making it maintainable and allowing us to deliver this performance improvement now. Additionally, it doesn't prevent a future migration to a unified system (once it is developed).

Nicky-2000 committed:
Improve comments as per the suggestions.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Remove static keyword on thread_local variable.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Add server.active_rdb_threads_num to ensure more
robust resource cleanup.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Make the rdb save worker buffer a config parameter.

This allows people to adjust the worker buffer
size based on their own workloads and memory
requirements.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Run rdb and replication tests with rdb-threads
set to different values.

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 committed:
Fix Spellcheck

Signed-off-by: Nicky Khorasani <nickykhorasani@google.com>

Nicky-2000 commented:
> > Overall LGTM. Some minor comments. Can we have some dedicated tcl for this feature?
> 
> Yep! I'll add a few tests that don't require the `--rdb-threads` flag in order to test the new functionality.

I modified integration/rdb.tcl, integration/replication-2.tcl and integration/replication-3.tcl to run in a loop with different number of rdb-threads enabled.
